getBoundingClientRect spends time flushing layout even if the current layout state is sufficiently valid.

mattwoodrow opened the pull request:
#### 2172f7b5f9f2ae8cd4ab6795400b6d3eb6c61f94
<pre>
getBoundingClientRect spends time flushing layout even if the current layout state is sufficiently valid.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287880">https://bugs.webkit.org/show_bug.cgi?id=287880</a>
&lt;<a href="https://rdar.apple.com/145082013">rdar://145082013</a>&gt;

Reviewed by Alan Baradlay.

Extends updateLayoutIfDimensionsOutOfDate to pass through LayoutOptions to the
final layout if required.

Adds Top and Left options, and rejects the fast path if any ancestor requires
position movement layout (on top of any ancestor requiring normal self layout).

Allows updateLayoutIfDimensionsOutOfDate to succeed if the queried renderer
requires only simplified layout, and it will succeed.

Start using updateLayoutIfDimensionsOutOfDate for getBoundingClientRect.

Moves a check from RenderBlock::simplifiedLayout() to
RenderBlock::canPerformSimplifiedLayout() (and make the latter public) so that
it correctly determines if the simplified layout will succeed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::boundingClientRect):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::canPerformSimplifiedLayout const):
(WebCore::RenderBlock::simplifiedLayout):
* Source/WebCore/rendering/RenderBlock.h:

Canonical link: <a href="https://commits.webkit.org/290922@main">https://commits.webkit.org/290922@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fe816dc0ff3afc7a6cfb6af5d90a5ae71eb15ea4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91464 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/10996 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/517 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96433 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42152 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93514 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11373 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19379 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/70233 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27748 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94465 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/8672 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/82857 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/50559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8437 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41322 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78757 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/449 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98436 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18626 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/13694 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/79255 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/18881 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78695 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/78459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19407 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/22977 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/343 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11761 "Failed to checkout and rebase branch from PR 41032") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18624 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/23900 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18334 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/21794 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20100 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3036,19 +3036,41 @@ bool Document::updateLayoutIfDimensionsOutOfDate(Element& element, OptionSet<Dim
+            }
+        }
+    }
+
```
Do you mind landing this in a dedicated commit? -in case of the unthinkable  'IgnoreOverflow ' regression, it would be bad to lose this fix too.

mattwoodrow replied:
“Unthinkable”, indeed.

None of the existing callers pass the flag to get this behaviour (maybe they should?), but happy to do it separately regardless.

mattwoodrow committed:
getBoundingClientRect spends time flushing layout even if the current layout state is sufficiently valid.
https://bugs.webkit.org/show_bug.cgi?id=287880
<rdar://145082013>

Reviewed by Alan Baradlay.

Extends updateLayoutIfDimensionsOutOfDate to pass through LayoutOptions to the
final layout if required.

Adds Top and Left options, and rejects the fast path if any ancestor requires
position movement layout (on top of any ancestor requiring normal self layout).

Allows updateLayoutIfDimensionsOutOfDate to succeed if the queried renderer
requires only simplified layout, and it will succeed.

Start using updateLayoutIfDimensionsOutOfDate for getBoundingClientRect.

Moves a check from RenderBlock::simplifiedLayout() to
RenderBlock::canPerformSimplifiedLayout() (and make the latter public) so that
it correctly determines if the simplified layout will succeed.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Document.h:
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::boundingClientRect):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::canPerformSimplifiedLayout const):
(WebCore::RenderBlock::simplifiedLayout):
* Source/WebCore/rendering/RenderBlock.h:

Canonical link: https://commits.webkit.org/290922@main
